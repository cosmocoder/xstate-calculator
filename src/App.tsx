import type { MouseEvent } from 'react';
import { Buttons } from './components/Buttons/Buttons';
import { Display } from './components/Display';
import type { ButtonData, MainOperations, OtherOperations } from './types';
import { useMachine } from '@xstate/react';
import { machine } from './machine';

export const App = () => {
  const [snapshot, send] = useMachine(machine);

  const handleClick = (e: MouseEvent<HTMLButtonElement>) => {
    const numFromClick = e.currentTarget.dataset.number;
    const opFromClick = e.currentTarget.dataset.operation as MainOperations;
    const otherFromClick = e.currentTarget.dataset.other as OtherOperations;

    if (numFromClick) {
      send({ type: 'number', payload: numFromClick });
      return;
    }

    if (opFromClick) {
      send({ type: 'operation', payload: opFromClick });
      return;
    }

    if (otherFromClick) {
      send({ type: 'other', payload: otherFromClick });
      return;
    }
  };

  const buttonData: ButtonData = {
    operation: snapshot.context.operation,
    clearOption:
      snapshot.context.currentNum ||
      snapshot.context.previousNum ||
      snapshot.context.result
        ? 'C'
        : 'AC',
  };

  return (
    <div className="calculator-container">
      <Display currentCalc={snapshot.context} />
      <Buttons handleClick={handleClick} buttonData={buttonData} />
    </div>
  );
};
