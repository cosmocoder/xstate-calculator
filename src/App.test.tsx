import { render, screen } from '@testing-library/react';
import { App } from './App';
import userEvent from '@testing-library/user-event';

describe('Calculator app', () => {
  it('should render correctly', () => {
    const { container } = render(<App />);

    expect(container).toMatchSnapshot();
  });

  it('should handle addition of two numbers', async () => {
    const user = userEvent.setup();
    render(<App />);

    await user.click(await screen.findByRole('button', { name: '1' }));
    await user.click(await screen.findByRole('button', { name: '+' }));
    await user.click(await screen.findByRole('button', { name: '2' }));
    await user.click(await screen.findByRole('button', { name: '=' }));

    expect(await screen.findByRole('region')).toHaveTextContent('3');
  });

  it('should handle subtraction of two numbers', async () => {
    const user = userEvent.setup();
    render(<App />);

    await user.click(await screen.findByRole('button', { name: '2' }));
    await user.click(await screen.findByRole('button', { name: '-' }));
    await user.click(await screen.findByRole('button', { name: '1' }));
    await user.click(await screen.findByRole('button', { name: '=' }));

    expect(await screen.findByRole('region')).toHaveTextContent('1');
  });

  it('should handle multiplication of two numbers', async () => {
    const user = userEvent.setup();
    render(<App />);

    await user.click(await screen.findByRole('button', { name: '2' }));
    await user.click(await screen.findByRole('button', { name: 'x' }));
    await user.click(await screen.findByRole('button', { name: '3' }));
    await user.click(await screen.findByRole('button', { name: '=' }));

    expect(await screen.findByRole('region')).toHaveTextContent('6');
  });

  it('should handle division of two numbers', async () => {
    const user = userEvent.setup();
    render(<App />);

    await user.click(await screen.findByRole('button', { name: '1' }));
    await user.click(await screen.findByRole('button', { name: 'รท' }));
    await user.click(await screen.findByRole('button', { name: '2' }));
    await user.click(await screen.findByRole('button', { name: '=' }));

    expect(await screen.findByRole('region')).toHaveTextContent('0.5');
  });

  it('should handle chaining of multiple math operations', async () => {
    const user = userEvent.setup();
    render(<App />);

    await user.click(await screen.findByRole('button', { name: '3' }));
    await user.click(await screen.findByRole('button', { name: '+' }));
    await user.click(await screen.findByRole('button', { name: '4' }));
    await user.click(await screen.findByRole('button', { name: '+' }));
    await user.click(await screen.findByRole('button', { name: '5' }));
    await user.click(await screen.findByRole('button', { name: '-' }));
    await user.click(await screen.findByRole('button', { name: '6' }));
    await user.click(await screen.findByRole('button', { name: 'รท' }));
    await user.click(await screen.findByRole('button', { name: '2' }));
    await user.click(await screen.findByRole('button', { name: '=' }));

    expect(await screen.findByRole('region')).toHaveTextContent('3');
  });

  it('should handle clearing of inputs', async () => {
    const user = userEvent.setup();
    render(<App />);

    expect(await screen.findByRole('button', { name: 'AC' })).toBeInTheDocument();

    await user.click(await screen.findByRole('button', { name: '3' }));
    expect(await screen.findByRole('button', { name: 'C' })).toBeInTheDocument();
    await user.click(await screen.findByRole('button', { name: 'C' }));

    expect(await screen.findByRole('region')).toHaveTextContent('0');

    await user.click(await screen.findByRole('button', { name: '2' }));
    await user.click(await screen.findByRole('button', { name: '+' }));
    await user.click(await screen.findByRole('button', { name: 'C' }));
    await user.click(await screen.findByRole('button', { name: '-' }));
    await user.click(await screen.findByRole('button', { name: '1' }));
    await user.click(await screen.findByRole('button', { name: '=' }));
    await user.click(await screen.findByRole('button', { name: '+' }));
    await user.click(await screen.findByRole('button', { name: 'C' }));
    await user.click(await screen.findByRole('button', { name: 'C' }));

    expect(await screen.findByRole('button', { name: 'AC' })).toBeInTheDocument();
    expect(await screen.findByRole('region')).toHaveTextContent('0');
  });

  it('should show decimal point correctly', async () => {
    const user = userEvent.setup();
    render(<App />);

    await user.click(await screen.findByRole('button', { name: '3' }));
    await user.click(await screen.findByRole('button', { name: '.' }));
    expect(await screen.findByRole('region')).toHaveTextContent('3.');
    await user.click(await screen.findByRole('button', { name: '2' }));
    expect(await screen.findByRole('region')).toHaveTextContent('3.2');

    await user.click(await screen.findByRole('button', { name: 'C' }));
    await user.click(await screen.findByRole('button', { name: '.' }));
    await user.click(await screen.findByRole('button', { name: '1' }));
    expect(await screen.findByRole('region')).toHaveTextContent('0.1');
  });

  it('should calculate percentage correctly', async () => {
    const user = userEvent.setup();
    render(<App />);

    await user.click(await screen.findByRole('button', { name: '2' }));
    await user.click(await screen.findByRole('button', { name: '%' }));
    expect(await screen.findByRole('region')).toHaveTextContent('0.02');

    await user.click(await screen.findByRole('button', { name: 'C' }));
    expect(await screen.findByRole('region')).toHaveTextContent('0');

    await user.click(await screen.findByRole('button', { name: '3' }));
    await user.click(await screen.findByRole('button', { name: '+' }));
    await user.click(await screen.findByRole('button', { name: '2' }));
    await user.click(await screen.findByRole('button', { name: '=' }));
    await user.click(await screen.findByRole('button', { name: '%' }));
    expect(await screen.findByRole('region')).toHaveTextContent('0.05');
  });

  it('should handle negation of numbers', async () => {
    const user = userEvent.setup();
    render(<App />);

    await user.click(await screen.findByRole('button', { name: '2' }));
    await user.click(await screen.findByRole('button', { name: '+/-' }));
    expect(await screen.findByRole('region')).toHaveTextContent('-2');

    await user.click(await screen.findByRole('button', { name: 'C' }));
    await user.click(await screen.findByRole('button', { name: '2' }));
    await user.click(await screen.findByRole('button', { name: '+' }));
    await user.click(await screen.findByRole('button', { name: '3' }));
    await user.click(await screen.findByRole('button', { name: '+/-' }));
    await user.click(await screen.findByRole('button', { name: '=' }));
    expect(await screen.findByRole('region')).toHaveTextContent('-1');

    await user.click(await screen.findByRole('button', { name: '%' }));
    await user.click(await screen.findByRole('button', { name: '+/-' }));
    expect(await screen.findByRole('region')).toHaveTextContent('0.01');
  });
});
