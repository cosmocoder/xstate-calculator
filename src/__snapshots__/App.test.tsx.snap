// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Calculator app > should render correctly 1`] = `
<div>
  <div
    class="calculator-container"
  >
    <div
      aria-live="polite"
      class="display"
      role="region"
    >
      0
    </div>
    <div
      class="buttons-container"
    >
      <button
        class="number-button zero "
        data-number="0"
        type="button"
      >
        0
      </button>
      <button
        class="number-button one "
        data-number="1"
        type="button"
      >
        1
      </button>
      <button
        class="number-button two "
        data-number="2"
        type="button"
      >
        2
      </button>
      <button
        class="number-button three "
        data-number="3"
        type="button"
      >
        3
      </button>
      <button
        class="number-button four "
        data-number="4"
        type="button"
      >
        4
      </button>
      <button
        class="number-button five "
        data-number="5"
        type="button"
      >
        5
      </button>
      <button
        class="number-button six "
        data-number="6"
        type="button"
      >
        6
      </button>
      <button
        class="number-button seven "
        data-number="7"
        type="button"
      >
        7
      </button>
      <button
        class="number-button eight "
        data-number="8"
        type="button"
      >
        8
      </button>
      <button
        class="number-button nine "
        data-number="9"
        type="button"
      >
        9
      </button>
      <button
        class="operation-button plus "
        data-operation="+"
        type="button"
      >
        +
      </button>
      <button
        class="operation-button minus "
        data-operation="-"
        type="button"
      >
        -
      </button>
      <button
        class="operation-button times "
        data-operation="x"
        type="button"
      >
        x
      </button>
      <button
        class="operation-button divide "
        data-operation="รท"
        type="button"
      >
        รท
      </button>
      <button
        class="button equals "
        data-other="="
        type="button"
      >
        =
      </button>
      <button
        class="button clear "
        data-other="clear"
        type="button"
      >
        AC
      </button>
      <button
        class="button decimal "
        data-other="."
        type="button"
      >
        .
      </button>
      <button
        class="button negative "
        data-other="+/-"
        type="button"
      >
        +/-
      </button>
      <button
        class="button percent "
        data-other="%"
        type="button"
      >
        %
      </button>
    </div>
  </div>
</div>
`;
